text/0000-forbidden-value-trait.md

```diff
@@ -0,0 +1,70 @@
+- Start Date: 2020-03-24
+- RFC PR #:
+- Rust Issue #:
+
+# Summary
+
+Some types have one or more forbidden values. Such values can be sometimes used to perform [optimizations](https://github.com/rust-lang/rust/pull/45225). For example, rustc currently uses a known forbidden value of a certain type `A` (like `&T`, `Box<T>`, `NonNull<T>`, `NonZeroI8` and some enum types) to represent the `None` enum variant of `Option<A>`. I propose to give users the ability to enable the same optimizations for types rustc currently doesn't know a forbidden value exists for.
+
+# Motivation
+
+This would increase the number of situations where the compiler can do the aforementioned, genuinely useful optimizations.
+
+# Detailed design
+
+Add the following trait somewhere in the standard library:
+```rust
+unsafe trait ForbiddenValues<A, B>
+where
+    A: FixedSizeArray<B>,
+    B: FixedSizeArray<u8>,
```

Comment:
Must this be `u8`?

Reply:
Yes, I think so. `B` is an array of bytes used for representing a forbidden bit pattern of `Self`. I believe that if `T` implements some `ForbiddenValues<A, B>`, then it could be undefined behaviour to have a value of type `T` with a bit pattern equal to one of the items in `FORBIDDEN_VALUES`.